### PR TITLE
`rustc_borrowck` cleanups, part 2

### PR DESCRIPTION
**Pull request by nnethercote**
The code under `do_mir_borrowck` is pretty messy, especially the various types like `MirBorrowckCtxt`, `BorrowckInferCtxt`, `MirTypeckResults`, `MirTypeckRegionConstraints`, `CreateResult`, `TypeChecker`, `TypeVerifier`, `LivenessContext`, `LivenessResults`. This PR does some tidying up, though there's still plenty of mess left afterwards.

A sequel to #132250.

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
I don't have the time or patience to review this right now

r? compiler

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 1]
```diff
@@ -2783,12 +2772,12 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
     ) -> ty::InstantiatedPredicates<'tcx> {
         if let Some(closure_requirements) = &tcx.mir_borrowck(def_id).closure_requirements {
             constraint_conversion::ConstraintConversion::new(
-                self.infcx,
+                self.infcx, // njn: change to BorrowckInferCtxt?
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/type_check/mod.rs:2775]: you should probably convert this to a fixme or address it

**Comment by Nadrieril** [comment]
All good for me, I checked the logic and the changes are pretty unambiguously improvements. Much appreciation for the small and well-explained commits.

I second Err's remark about the FIXME comment; r=me once that is fixed if perf is good.

**Comment by nnethercote** [comment]
I think the perf results are just noise.

@rustbot label: +perf-regression-triaged

**Comment by nnethercote** [comment]
Perf is ok and I have rebased.

@bors r=Nadrieril

**Comment by nnethercote** [comment]
I rebased.

@bors r=Nadrieril

**Comment by nnethercote** [comment]
I did some local measurements that strongly indicated that the final commit was responsible for the wg-grammar regression. So I filed #133220 to revert that commit, but the perf run there showed it had no effect. Weird. I don't think it's worth spending more time on a mysterious regression that affects just a single secondary benchmark.

**Pull request merged**